templates/openstackconfiggenerator/bin/process-roles.py

```diff
@@ -0,0 +1,167 @@
+#! /usr/bin/env python3
+
+import argparse
+import collections
+import logging
+import sys
+import unittest
+import unittest.mock
+import yaml
+
+logging.basicConfig(stream=sys.stdout, level=logging.INFO)
+LOG = logging.getLogger('process-roles')
+
+
+class TemplateLoader(yaml.SafeLoader):
+    def construct_mapping(self, node):
+        self.flatten_mapping(node)
+        return collections.OrderedDict(self.construct_pairs(node))
+
+
+class TemplateDumper(yaml.SafeDumper):
+    def represent_ordered_dict(self, data):
+        return self.represent_dict(data.items())
+
+    def description_presenter(self, data):
+        if not len(data) > 80:
+            return self.represent_scalar('tag:yaml.org,2002:str', data)
+        return self.represent_scalar('tag:yaml.org,2002:str', data, style='>')
+
+
+TemplateDumper.add_representer(str,
+                               TemplateDumper.description_presenter)
+TemplateDumper.add_representer(collections.OrderedDict,
+                               TemplateDumper.represent_ordered_dict)
+TemplateLoader.add_constructor(yaml.resolver.BaseResolver.DEFAULT_MAPPING_TAG,
+                               TemplateLoader.construct_mapping)
+
+
+def get_role_counts(environment):
+    """Find role counts in heat environment files"""
+    role_counts = {}
+    for env_file in environment:
+        try:
+            with open(env_file, 'r', encoding='utf-8') as envfile:
+                data = yaml.safe_load(envfile.read())
+            if data is not None:
+                for param, val in data.get('parameter_defaults', {}).items():
+                    if param.endswith('Count'):
```

Comment:
from a quick grep there are some parameters (DockerPuppetProcessCount, CinderXtremioArrayBusyRetryCount, IronicDhcpv6StatefulAddressCount, KeystoneUniqueLastPasswordCount, CollectdSensubilityWorkerCount) which also end with Count. There is a chance we get those added to role_counts. we could first get all role names from roles_file and pass them here.

Reply:
Yes, false matches were a concern. If it's not an integer then we ignore it completely. Keeping these params that look like a role count should not be an issue since we only lookup the counts for the roles defined in roles_data